File: ZPZAppleFrameWorkPractice/OC_FrameWork/AddressBook_Contact_FrameWork/ZPZ_AB_CN_DataViewController.m
//
//  ZPZ_AB_CN_DataViewController.m
//  ZPZAppleFrameWork
//
//  Created by zhoupengzu on 2017/9/25.
//  Copyright © 2017年 zhoupengzu. All rights reserved.
//

#import "ZPZ_AB_CN_DataViewController.h"
#import <Contacts/Contacts.h>
#import <AddressBook/AddressBook.h>
#import "ZPZCommonMethod.h"

@interface ZPZ_AB_CN_DataViewController ()

@end

@implementation ZPZ_AB_CN_DataViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.edgesForExtendedLayout = UIRectEdgeNone;
    self.view.backgroundColor = [UIColor whiteColor];
    UIButton * readButton = [UIButton buttonWithType:UIButtonTypeCustom];
    readButton.backgroundColor = [UIColor orangeColor];
    readButton.frame = CGRectMake(20, 20, self.view.frame.size.width - 2 * 20, 50);
    readButton.layer.cornerRadius = readButton.frame.size.height / 2;
    readButton.layer.masksToBounds = YES;
    [readButton setTitle:@"read" forState:UIControlStateNormal];
    [readButton addTarget:self action:@selector(clickedReadLocalData) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:readButton];
}

- (void)clickedReadLocalData{
    if ([ZPZCommonMethod isNeedNewContactFrameWork]) {
        CNAuthorizationStatus status = [CNContactStore authorizationStatusForEntityType:CNEntityTypeContacts];
        switch (status) {
            case CNAuthorizationStatusNotDetermined:{
                CNContactStore * store = [[CNContactStore alloc] init];
                [store requestAccessForEntityType:CNEntityTypeContacts completionHandler:^(BOOL granted, NSError * _Nullable error) {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        if (granted) {
                            [self gotoReadData];
                        }else{
                            [self showNotAllowedTip];
                        }
                    });
                }];
            }
                break;
            case CNAuthorizationStatusAuthorized:{
                [self gotoReadData];
            }
                break;
            case CNAuthorizationStatusDenied:
            case CNAuthorizationStatusRestricted:{
                [self showNotAllowedTip];
            }
                break;
            default:
                break;
        }
    }else{
        ABAuthorizationStatus status = ABAddressBookGetAuthorizationStatus();
        switch (status) {
            case kABAuthorizationStatusNotDetermined:
            {
                ABAddressBookRef addressBook = ABAddressBookCreate();
                ABAddressBookRequestAccessWithCompletion(addressBook, ^(bool granted, CFErrorRef error) {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        if (granted) {
                            [self gotoReadData];
                        }else{
                            [self showNotAllowedTip];
                        }
                    });
                });
                CFRelease(addressBook);
            }
                break;
            case kABAuthorizationStatusAuthorized:{
                [self gotoReadData];
            }
                break;
            case kABAuthorizationStatusDenied:
            case kABAuthorizationStatusRestricted:{
                [self showNotAllowedTip];
            }
            default:
                break;
        }
    }
}

- (void)gotoReadData{
    if ([ZPZCommonMethod isNeedNewContactFrameWork]) {
        CNContactStore * store = [[CNContactStore alloc] init];
        CNContactFetchRequest * request = [[CNContactFetchRequest alloc] initWithKeysToFetch:@[CNContactGivenNameKey,CNContactNicknameKey,CNContactPhoneNumbersKey]];
        BOOL isEnume = [store enumerateContactsWithFetchRequest:request error:NULL usingBlock:^(CNContact * _Nonnull contact, BOOL * _Nonnull stop) {
            NSLog(@"phone:%@",contact.phoneNumbers);
        }];
        if (!isEnume) {
            NSLog(@"read error");
        }
    }else{
        ABAddressBookRef addressBook = ABAddressBookCreate();
        CFArrayRef recordArr = ABAddressBookCopyArrayOfAllPeople(addressBook);
        for (CFIndex i = 0; i < CFArrayGetCount(recordArr); i++) {
            ABRecordRef multiValue = CFArrayGetValueAtIndex(recordArr, i);
            NSLog(@"%@",ABRecordCopyValue(multiValue, kABPersonLastNameProperty));
        }
        NSLog(@"%@",recordArr);
        CFRelease(recordArr);
        CFRelease(addressBook);
    }
}

- (void)showNotAllowedTip{
    UIAlertController * alertVC = [UIAlertController alertControllerWithTitle:@"提示" message:@"您关闭了我们的权限，您可以去设置中去手动打开" preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction * knowAction = [UIAlertAction actionWithTitle:@"知道了" style:UIAlertActionStyleCancel handler:nil];
    if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]]) {
        UIAlertAction * goAction = [UIAlertAction actionWithTitle:@"去打开" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]];
        }];
        [alertVC addAction:goAction];
    }
    [alertVC addAction:knowAction];
    [self presentViewController:alertVC animated:YES completion:nil];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: ZPZAppleFrameWorkPractice/Swift_FrameWork/AddressBook_Contact_FrameWork/ZPZ_AB_CN_DataViewController.swift
//
//  ZPZ_AB_CN_DataViewController.swift
//  ZPZAppleFrameWork
//
//  Created by zhoupengzu on 2017/9/21.
//  Copyright © 2017年 zhoupengzu. All rights reserved.
//

import UIKit
import Contacts
import AddressBook

class ZPZ_AB_CN_DataViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        self.edgesForExtendedLayout = UIRectEdge.init(rawValue: 0)
        self.view.backgroundColor = UIColor.white
        configUI()
    }
    func configUI(){
        let button: UIButton = UIButton.init(type: .custom)
        button.backgroundColor = UIColor.orange
        button.frame = CGRect(x: 20, y: 20, width: view.frame.size.width - 2 * 20, height: 50)
        button.layer.cornerRadius = button.frame.size.height / 2
        button.layer.masksToBounds = true
        button.setTitle("开始获取", for: .normal)
        button.addTarget(self, action: #selector(gotoOpenSystemAddressBook), for: .touchUpInside)
        view .addSubview(button)
    }
    
    @objc func gotoOpenSystemAddressBook(){
//        if #available(iOS 9.0, *) {
//            let status = CNContactStore.authorizationStatus(for: .contacts)
//            switch status{
//            case .notDetermined:
//                let contactStore = CNContactStore()
//                contactStore.requestAccess(for: CNEntityType.contacts, completionHandler: { (isAllow, error) in
//                    DispatchQueue.main.async {
//                        if isAllow{
//                            self.gotoReadData()
//                        }else{
//                            self.showHaveNoPermissionTip()
//                        }
//                    }
//                })
//            case .denied:
//                fallthrough
//            case .restricted:
//                showHaveNoPermissionTip()
//            case .authorized:
//                gotoReadData()
//            }
//        }else{
            let status = ABAddressBookGetAuthorizationStatus()
            switch status{
            case .notDetermined:
                let addressBook = ABAddressBookCreate().takeRetainedValue()
                ABAddressBookRequestAccessWithCompletion(addressBook, { (isAllow, error) in
                    DispatchQueue.main.async {
                        if isAllow{
                            self.gotoReadData()
                        }else{
                            self.showHaveNoPermissionTip()
                        }
                    }
                })
            case .denied,.restricted:
                showHaveNoPermissionTip()
            case .authorized:
                gotoReadData()
            }
//        }
    }
    
    func gotoReadData(){
        if #available(iOS 9.0, *){
            let contactRequest = CNContactFetchRequest.init(keysToFetch: [CNContactPhoneNumbersKey,CNContactNamePrefixKey,CNContactGivenNameKey] as [CNKeyDescriptor])
            let contactStore = CNContactStore.init()

            do{
                try contactStore.enumerateContacts(with: contactRequest, usingBlock: { (contact, stop) in
                    ZPZContactGetInfo.printContactInfo(baseInfo: contact)
                })
            }catch let error as NSError{
                print(error)
            }
        }else{
            let addressBook = ABAddressBookCreate().takeRetainedValue()
            let allPeople:NSArray = ABAddressBookCopyArrayOfAllPeople(addressBook).takeRetainedValue() as NSArray
            for i in 0..<allPeople.count{
                let person = allPeople[i]
                ZPZAddressBookGetInfo.printAddressBookInfo(baseInfo: person as ABRecord)
            }
            //下面这么写有问题？？？？？
//            let addressBook = ABAddressBookCreate().takeRetainedValue()
//            let allPeople:CFArray = ABAddressBookCopyArrayOfAllPeople(addressBook).takeRetainedValue()
//            for i in 0..<CFArrayGetCount(allPeople){
//                let person = CFArrayGetValueAtIndex(allPeople, i)
//                ZPZAddressBookGetInfo.printAddressBookInfo(baseInfo: person as ABRecord)
//            }
        }
    }
    /**
     CFArrayRef peopleArr = ABAddressBookCopyArrayOfAllPeople(ABAddressBookCreate());
     NSLog(@"%@",peopleArr);
     for (NSInteger i = 0; i < CFArrayGetCount(peopleArr); i++) {
     NSLog(@"%@",ABRecordCopyValue(CFArrayGetValueAtIndex(peopleArr, i), kABPersonFirstNameProperty));
     }
     */
    func showHaveNoPermissionTip(){
        let alertVC = UIAlertController.init(title: "提示", message: "您已经拒绝了我们访问您的通讯录，您可以自己去设置里面打开", preferredStyle: .alert)
        let knowAction = UIAlertAction.init(title: "知道了", style: .cancel, handler: nil)
        let openUrl = URL.init(string: UIApplicationOpenSettingsURLString)
        if let tempOpenUrl = openUrl{
            let canOpenSetting = UIApplication.shared.canOpenURL(tempOpenUrl)
            if canOpenSetting {
                let goAction = UIAlertAction.init(title: "去设置", style: .default, handler: { (go) in
                    UIApplication.shared.openURL(tempOpenUrl)
                })
                alertVC.addAction(goAction)
            }
        }
        alertVC.addAction(knowAction)
        self.present(alertVC, animated: true, completion: nil)
    }
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

}
